keps/sig-network/4427-relaxed-dns-search-validation/README.md

```diff
@@ -0,0 +1,813 @@
+      - 1.2.3.4
+    searches:
+      - abc_d.example.com
+```
+
+However, this returns an error:
+
+```
+The Pod "dns-example" is invalid: spec.dnsConfig.searches[0]: Invalid value: "abc_d.example.com": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
+```
+
+While a case could be made that these names are incorrect as host names based on RFC-1123, DNS servers generally allow such names
+and once they are in use DNS names can be complicated to change especially when they're used by legacy systems and not purely
+within Kubernetes. The DNS searches in the Kubernetes pods spec are not defining the host name, so even under the stricter interpretation
+of RFC-1123, relaxing these checks does not allow users of Kubernetes to introduce an invalid host name where one did not
+already exist outside of Kubernetes.
+
+### Goals
+
+- Support workloads that need to resolve DNS short names where the full DNS name includes characters disallowed by RFC-1123.
+
+## Proposal
+
+Introduce a RelaxedDNSSearchValidation feature gate which is disabled by default. When the feature gate is enabled,
+DNS search strings will be validated according to RFC-2181:
+- The length of any one label is limited to between 1 and 63 octets
+- A full domain name is limited to 255 octets (including the separators)
+
+Since the relaxed check allows previously invalid values, care must be taken to support cluster downgrades safely. To accomplish this, the validation will distinguish between new resources and updates to existing resources:
+-	When the feature gate is disabled:
+  - (a) New resources will use strict validation based on RFC-1123 (no change to current validation)
+  - (b) Updates to existing resources will use relaxed validation if any search string in the existing list fails strict validation
+- When the feature gate is enabled:
+  - (c) New resources will use relaxed validation.
+  - (d) Updates to existing resources will use relaxed validation.
+
+This means that it is safe to downgrade a cluster with the feature gate enabled to a version where the feature gate is present (whether it’s enabled or disabled). It is not safe, in general, to downgrade from a cluster with the gate enabled to a version prior to the gate being introduced, since values may have been written to storage which will no longer pass validation. However, this scenario requires opting in through enabling the gate. In practice, the recommended approach would be to only enable to the gate after upgrading from a version with relaxed checking already present.
+
+As long as the gate is disabled, there is no compatibility change, so cluster downgrades are not affected by the feature.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
```

Comment:
Are there no risks worth calling out?  

Are we 100% confident this won't break some app's resolver library?  That seems like a (minor) risk.

You could call out that all the risk of this feature is borne by pods which opt in to it - apps might break, but they control their own settings.  This should have no risk to the cluster, the control plane, the kubelet, etc.

Right?

Reply:
Added a note on the opt-in nature. Risk of confusing a resolver (or YAML parsing) is eliminated by restricting this to just adding underscores.